activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -507,6 +507,7 @@ def checkout(checkout_timeout = @checkout_timeout)
       # +conn+: an AbstractAdapter object, which was obtained by earlier by
       # calling #checkout on this pool.
       def checkin(conn)
+        conn.assert_connection_reuseable
```

Comment:
I assumed this only happens in case of progammer error and the earlier we blow up the better.

I can think of 2 cases this can happen:

1. Calling checkin inside a transaction:

    ```ruby
    Post.transaction do
        ...
        AR::connection.close # calls poll.checkin connection
    end # [*]
    ```
2. incorrect transaction use, not comiting or rollingback called after begin_transaction

    ```ruby
       AR::Base.connection.begin_transaction
       ...
       AR::Base.connection.close
    ```

If we roll back on case 1 it'll blow up when the transaction block original tries to release the transaction at [*], as it's already released. That should do as well, but will be a bit harder to see that where the error is coming from.
In case 2 rolling back automatically kind of silents the issue. So maybe it's a good idea to add an error log as well.
 
In both case i think the last thing we want is the current behaviour of happily allowing reusing the connection with an opened transaction.

I guess mysql close and reconnect should be fine as we'd be calling reset! only if we have open transactions.